### ISSUE TITLE
Fail in rendering of feedback text

### ISSUE THREAD

**Issue by mkraska** [maths/moodle-qtype_stack#829]
I encounter a problem new in 4.4 with display of feedback text:

The content of the feedback cell is (HTML format):

    \(Q \) für  \(x /L\) {@text@}

It displays as:

![image](https://user-images.githubusercontent.com/12987918/182952116-bec1e001-f032-43ce-a180-7e786c057cb6.png)

The text with Q and x/L is converted into this strange `//CT2S21//` whereas the variable `text` is displayed correctly.

Any idea? 

[quiz-T-B1-MB-TM1-TM1 09 05 Schnittlastenverlauf 1-20220804-2309.zip](https://github.com/maths/moodle-qtype_stack/files/9263361/quiz-T-B1-MB-TM1-TM1.09.05.Schnittlastenverlauf.1-20220804-2309.zip)

**Comment by aharjula** [comment]
Those strange things are placeholders for static content that was extracted from the CASText before the CASText was sent to the CAS to be evaluated and are supposed to be replaced with that content after the content comes back from the CAS. This is a feature that has been seen working and and should work but testing now says that it does not so I will need to do some searching.

The feature is especially important for your use cases as it is the thing that stop us from sending those massive static JavaScript segments to the CAS and to the CAS-cache where they would eat bandwidth and storage, and slow things down. In this case the replaced string just happens to be on the sorter end of the string the logic targets.

The good thing here is that the questions compiled data structure does contain that string (`"\/\/CT2S21\/\/":" \\(Q \\) f\u00fcr  \\(x \/L\\) "`, this particular question has a total of 25 such strings and they are not all that short, all of them are stored within a JSON dictionary in the `compiledcache` column of the qtype_stack_options table) so it has been correctly collected and is in the correct place so the issue is only with the bit of logic that is supposed to return it into its place. There will be a patch and it will again mean some work for your technical support as it will require some changes to the PHP scripts.

**Comment by mkraska** [comment]
Thanks for this explanation. It points me to a workaround: add dynamic content to the line:

    {@'a@} für  \(t / \mathrm{s} \) {@text@}

![image](https://user-images.githubusercontent.com/12987918/183021163-23cfee5d-ff9c-4ff3-9d16-8aba2289a887.png)

**Comment by mkraska** [comment]
Another question resulting from your explanation:

What is the scope of string caching? In the above example the injections prevent caching of the string inbetween (or enforce correct display)

My current textblock to include meclib images is

```
[[jsxgraph width='250px' height='250px' ]] 
var mode  = "STACK";
var stateRef;
const initstring = {#init#};
const centeredLabelStyle = {size:0, showInfobox:false, label:{offset:[-6,0], 
  anchorX:'left', anchorY:'middle'}};
// End of STACK header
[[include src="https://raw.githubusercontent.com/mkraska/meclib/main/meclib.js" /]]
[[/jsxgraph]]
```
So there is a continuous block of text with just one injection of the variable `init`. Will that prevent smart caching of the included stuff?

**Comment by aharjula** [comment]
That dynamic content is all that is required to trigger full post-processing to take place, and apparently I had not correctly written the simpler branch of logic that deals with content that does not require post-processing but should still get those static portions returned to their correct places.

Anyway, the above commit is all that neds to be changed for this to work correctly.

**Comment by aharjula** [comment]
No need to be continuous. The static string extraction works on the compiled CASText2 statement, which is essentially a list of string values separated with sections that turn injected values into strings. The logic goes through that list and identifies those strings that are just strings without any special role and are long enough and replaces them with those placeholders, there is no limit on the number of such replacements in a single CASText content (e.g. question-text), and you don't need to worry about placing injections within continuous content.

As to caching, this mainly saves room in the cache the injected value will naturally result in something that varies and still leads to more items in the cache we aim to remove the bits that would not vary in the thing that gets cached and only store those in a single place.

In your examples case the compiled CASText2 would look like this, with some variation in the numeric indetifiers:
```
[
  "%root", 
  [
    "jsxgraph", 
    "//CT2S0//", 
    "//CT2S1//",
    block([_ct2_tmp], _EC(errcatch(_ct2_tmp : init), "test/4:20-4:28"), string(_ct2_tmp)),
    "//CS2T2//"
    ]
]
```
Basically, a single JSXGraph block with its static payload separated from everything else. In the case of JSXGraph the first placeholder holds a JSON object that would contain the size parameters (and any other it might have that are not being used inside the CAS) of that graph as a separate entity, but the ones following that represent the content and various error trapping logic for the injection.

That would be what would go into the CAS and pretty much the same would come out but with that `init` having been evaluated and turned into a string the whole list would then get post-processed and the JSXGraph plugged into the JavaScript system of Moodle.

**Comment by mkraska** [comment]
Fix applied, problem solved, thanks a lot.
